### PR TITLE
Publish snapshots to maven via GHA

### LINKED ISSUE #620: Publish snapshots to maven via GHA
Coming from [opensearch-build#3185](https://github.com/opensearch-project/opensearch-build/issues/3185)

### What is happening?

We are de-coupling the task of publishing the maven snapshots from centralized build workflow to individual repositories. What this means is each repo can now publish maven snapshots using GitHub Actions.
This change unblocks the dependent components from waiting for a successful build before they can consume the snapshots. This also ensures that all snapshots are independent and up to date.

### What do you need to do?
1. Determine what artifacts you publish to maven. Most plugins publish only zips. [Example](https://aws.oss.sonatype.org/content/repositories/snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/).
2. Add below repository under `publishing` in build.gradle:

```
    repositories {
        maven {
            name = "Snapshots"
            url = "https://aws.oss.sonatype.org/content/repositories/snapshots"
            credentials {
                username "$System.env.SONATYPE_USERNAME"
                password "$System.env.SONATYPE_PASSWORD"
            }
        }
    }
```
[Example](https://github.com/opensearch-project/security/blob/main/build.gradle#L233-L242)

3. Determine what gradle task will publish the content to snapshots repo. Example: `./gradlew publishPluginZipPublicationToSnapshotsRepository`.
4. Create a pull request to set up a GitHub action that will fetch the credentials and publish the snpahots to https://aws.oss.sonatype.org/ [See sample workflow](https://github.com/opensearch-project/security/blob/main/.github/workflows/maven-publish.yml).
5. Add @gaiksaya as a reviewer, who will also take care of creating the IAM role that is required to fetch the credentials and add it to the GitHub Secrets.
### Testing

Test by editing the snapshot repository by updating the Snapshots target repository from maven to a local file system path.

build.gradle
```
        maven {
            name = "Snapshots" //  optional target repository name
            url = "https://aws.oss.sonatype.org/content/repositories/snapshots"
            credentials {
                username "$System.env.SONATYPE_USERNAME"
                password "$System.env.SONATYPE_PASSWORD"
            }
        }
```
->
```
        maven {
            name = "Snapshots" //  optional target repository name
            url = "snapshots"
        }
```
Run the command ./gradlew <taskName> which will publish the files in the file system under the 'snapshots' folder, then scan with `find` or your favor directory scanner to see the zip + pom with sha files

```
% find snapshots | sort                                                                                                                                                                               ~/git/security
snapshots
snapshots/org
snapshots/org/opensearch
snapshots/org/opensearch/plugin
snapshots/org/opensearch/plugin/opensearch-security
snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT
snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/maven-metadata.xml
snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/maven-metadata.xml.md5
snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/maven-metadata.xml.sha1
snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/maven-metadata.xml.sha256
snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/maven-metadata.xml.sha512
snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/opensearch-security-3.0.0.0-20230214.224249-1.pom
snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/opensearch-security-3.0.0.0-20230214.224249-1.pom.md5
snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/opensearch-security-3.0.0.0-20230214.224249-1.pom.sha1
snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/opensearch-security-3.0.0.0-20230214.224249-1.pom.sha256
snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/opensearch-security-3.0.0.0-20230214.224249-1.pom.sha512
snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/opensearch-security-3.0.0.0-20230214.224249-1.zip
snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/opensearch-security-3.0.0.0-20230214.224249-1.zip.md5
snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/opensearch-security-3.0.0.0-20230214.224249-1.zip.sha1
snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/opensearch-security-3.0.0.0-20230214.224249-1.zip.sha256
snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/opensearch-security-3.0.0.0-20230214.224249-1.zip.sha512
snapshots/org/opensearch/plugin/opensearch-security/maven-metadata.xml
snapshots/org/opensearch/plugin/opensearch-security/maven-metadata.xml.md5
snapshots/org/opensearch/plugin/opensearch-security/maven-metadata.xml.sha1
snapshots/org/opensearch/plugin/opensearch-security/maven-metadata.xml.sha256
snapshots/org/opensearch/plugin/opensearch-security/maven-metadata.xml.sha512
```

Please feel free to reach out to @opensearch-project/engineering-effectiveness.

### PR DESCRIPTION
**Pull request by lezzago**
### Description
Publish snapshots to maven via GHA

### Issues Resolved
#620

### Testing:
```
find notifications/snapshots | sort
notifications/snapshots
notifications/snapshots/org
notifications/snapshots/org/opensearch
notifications/snapshots/org/opensearch/plugin
notifications/snapshots/org/opensearch/plugin/notifications
notifications/snapshots/org/opensearch/plugin/notifications/3.0.0.0-SNAPSHOT
notifications/snapshots/org/opensearch/plugin/notifications/3.0.0.0-SNAPSHOT/maven-metadata.xml
notifications/snapshots/org/opensearch/plugin/notifications/3.0.0.0-SNAPSHOT/maven-metadata.xml.md5
notifications/snapshots/org/opensearch/plugin/notifications/3.0.0.0-SNAPSHOT/maven-metadata.xml.sha1
notifications/snapshots/org/opensearch/plugin/notifications/3.0.0.0-SNAPSHOT/maven-metadata.xml.sha256
notifications/snapshots/org/opensearch/plugin/notifications/3.0.0.0-SNAPSHOT/maven-metadata.xml.sha512
notifications/snapshots/org/opensearch/plugin/notifications/3.0.0.0-SNAPSHOT/notifications-3.0.0.0-20230221.175731-1.pom
notifications/snapshots/org/opensearch/plugin/notifications/3.0.0.0-SNAPSHOT/notifications-3.0.0.0-20230221.175731-1.pom.md5
notifications/snapshots/org/opensearch/plugin/notifications/3.0.0.0-SNAPSHOT/notifications-3.0.0.0-20230221.175731-1.pom.sha1
notifications/snapshots/org/opensearch/plugin/notifications/3.0.0.0-SNAPSHOT/notifications-3.0.0.0-20230221.175731-1.pom.sha256
notifications/snapshots/org/opensearch/plugin/notifications/3.0.0.0-SNAPSHOT/notifications-3.0.0.0-20230221.175731-1.pom.sha512
notifications/snapshots/org/opensearch/plugin/notifications/3.0.0.0-SNAPSHOT/notifications-3.0.0.0-20230221.175731-1.zip
notifications/snapshots/org/opensearch/plugin/notifications/3.0.0.0-SNAPSHOT/notifications-3.0.0.0-20230221.175731-1.zip.md5
notifications/snapshots/org/opensearch/plugin/notifications/3.0.0.0-SNAPSHOT/notifications-3.0.0.0-20230221.175731-1.zip.sha1
notifications/snapshots/org/opensearch/plugin/notifications/3.0.0.0-SNAPSHOT/notifications-3.0.0.0-20230221.175731-1.zip.sha256
notifications/snapshots/org/opensearch/plugin/notifications/3.0.0.0-SNAPSHOT/notifications-3.0.0.0-20230221.175731-1.zip.sha512
notifications/snapshots/org/opensearch/plugin/notifications/maven-metadata.xml
notifications/snapshots/org/opensearch/plugin/notifications/maven-metadata.xml.md5
notifications/snapshots/org/opensearch/plugin/notifications/maven-metadata.xml.sha1
notifications/snapshots/org/opensearch/plugin/notifications/maven-metadata.xml.sha256
notifications/snapshots/org/opensearch/plugin/notifications/maven-metadata.xml.sha512
```

```
find core/snapshots | sort
core/snapshots
core/snapshots/org
core/snapshots/org/opensearch
core/snapshots/org/opensearch/plugin
core/snapshots/org/opensearch/plugin/opensearch-notifications-core
core/snapshots/org/opensearch/plugin/opensearch-notifications-core/3.0.0.0-SNAPSHOT
core/snapshots/org/opensearch/plugin/opensearch-notifications-core/3.0.0.0-SNAPSHOT/maven-metadata.xml
core/snapshots/org/opensearch/plugin/opensearch-notifications-core/3.0.0.0-SNAPSHOT/maven-metadata.xml.md5
core/snapshots/org/opensearch/plugin/opensearch-notifications-core/3.0.0.0-SNAPSHOT/maven-metadata.xml.sha1
core/snapshots/org/opensearch/plugin/opensearch-notifications-core/3.0.0.0-SNAPSHOT/maven-metadata.xml.sha256
core/snapshots/org/opensearch/plugin/opensearch-notifications-core/3.0.0.0-SNAPSHOT/maven-metadata.xml.sha512
core/snapshots/org/opensearch/plugin/opensearch-notifications-core/3.0.0.0-SNAPSHOT/opensearch-notifications-core-3.0.0.0-20230221.175731-1.pom
core/snapshots/org/opensearch/plugin/opensearch-notifications-core/3.0.0.0-SNAPSHOT/opensearch-notifications-core-3.0.0.0-20230221.175731-1.pom.md5
core/snapshots/org/opensearch/plugin/opensearch-notifications-core/3.0.0.0-SNAPSHOT/opensearch-notifications-core-3.0.0.0-20230221.175731-1.pom.sha1
core/snapshots/org/opensearch/plugin/opensearch-notifications-core/3.0.0.0-SNAPSHOT/opensearch-notifications-core-3.0.0.0-20230221.175731-1.pom.sha256
core/snapshots/org/opensearch/plugin/opensearch-notifications-core/3.0.0.0-SNAPSHOT/opensearch-notifications-core-3.0.0.0-20230221.175731-1.pom.sha512
core/snapshots/org/opensearch/plugin/opensearch-notifications-core/3.0.0.0-SNAPSHOT/opensearch-notifications-core-3.0.0.0-20230221.175731-1.zip
core/snapshots/org/opensearch/plugin/opensearch-notifications-core/3.0.0.0-SNAPSHOT/opensearch-notifications-core-3.0.0.0-20230221.175731-1.zip.md5
core/snapshots/org/opensearch/plugin/opensearch-notifications-core/3.0.0.0-SNAPSHOT/opensearch-notifications-core-3.0.0.0-20230221.175731-1.zip.sha1
core/snapshots/org/opensearch/plugin/opensearch-notifications-core/3.0.0.0-SNAPSHOT/opensearch-notifications-core-3.0.0.0-20230221.175731-1.zip.sha256
core/snapshots/org/opensearch/plugin/opensearch-notifications-core/3.0.0.0-SNAPSHOT/opensearch-notifications-core-3.0.0.0-20230221.175731-1.zip.sha512
core/snapshots/org/opensearch/plugin/opensearch-notifications-core/maven-metadata.xml
core/snapshots/org/opensearch/plugin/opensearch-notifications-core/maven-metadata.xml.md5
core/snapshots/org/opensearch/plugin/opensearch-notifications-core/maven-metadata.xml.sha1
core/snapshots/org/opensearch/plugin/opensearch-notifications-core/maven-metadata.xml.sha256
core/snapshots/org/opensearch/plugin/opensearch-notifications-core/maven-metadata.xml.sha512
```

### Check List
- [X] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: .github/workflows/maven-publish.yml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        2.*
+      ]
+
+jobs:
+  build-and-publish-snapshots:
+    strategy:
+      fail-fast: false
+      matrix:
+        jdk: [11, 17]
+        platform: ["ubuntu-latest", "windows-latest", "macos-latest"]
```
Comment (gaiksaya) [inline .github/workflows/maven-publish.yml:18]: I dont think we need to build on all platforms to publish to maven. Just `ubuntu-latest` works

**Review by gaiksaya — COMMENTED** [review]
Can you post the out of all the publications in the comment just to confirm? Testing information is provided in the issue
Thanks!

File: .github/workflows/maven-publish.yml (hunk 1 above)
Comment (lezzago) [reply .github/workflows/maven-publish.yml]: done

File: .github/workflows/maven-publish.yml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        2.*
+      ]
+
+jobs:
+  build-and-publish-snapshots:
+    strategy:
+      fail-fast: false
+      matrix:
+        jdk: [17]
+        platform: ["ubuntu-latest"]
```
Comment (gaiksaya) [inline .github/workflows/maven-publish.yml:18]: Nit: Since it is just one value you can remove the matrix.

**Comment by gaiksaya** [comment]
Added the role in secrets! Thanks!

**Review by getsaurabh02 — DISMISSED** [review]

**Review by gaiksaya — APPROVED** [review]

**Review by getsaurabh02 — APPROVED** [review]

**Pull request merged**